src/main/java/pl/zaradny/springApp/infrastructure/InMemoryProductRepository.java

```diff
@@ -0,0 +1,54 @@
+package pl.zaradny.springApp.infrastructure;
+
+import org.springframework.stereotype.Repository;
+import pl.zaradny.springApp.domain.Product;
+import pl.zaradny.springApp.exceptions.ProductNotFoundException;
+
+import java.util.HashMap;
+import java.util.Map;
+
+@Repository
+public class InMemoryProductRepository implements ProductRepository {
+
+    private final Map<String, Product> products = new HashMap<>();
+
+    @Override
+    public void save(Product product) {
+        products.put(product.getId(), product);
+    }
+
+    @Override
+    public Product findById(String id) {
+
+        if(products.containsKey(id)){
+            return products.get(id);
+        }
+
+        throw new ProductNotFoundException();
+    }
+
+    @Override
+    public void deleteById(String id) {
+
+        if(products.containsKey(id)){
+            products.remove(id);
+        }else{
+            throw new ProductNotFoundException();
+        }
+
+    }
+
+    @Override
+    public Product updateById(Product product, String name) {
+
+        if(products.containsKey(product.getId())){
```

Comment:
Zwróć uwagę, że wykonujesz double-check (na poziomie API wyciągasz produkt z bazy, a tutaj sprawdzasz jeszcze raz czy na pewno tam jest). W naszym przypadku nie stanowi to problemu, bo baza jest w pamięci, ale w przypadku normalnej bazy danych wygenerowałbyś tam 2 żądania (o ile nie miałbyś cache). Nie musi być to coś złego, ale miej na względzie, że czasami może być to problematyczne :)